src/Razor/test/Microsoft.VisualStudio.LanguageServerClient.Razor.Test/RazorProjectInfoPublisherTest.cs

```diff
@@ -193,7 +198,9 @@ public async Task ProjectManager_Changed_DocumentOpened_InitializedProject_NoFil
         });
 
         // Assert
-        Assert.Empty(publisher.DeferredPublishTasks);
+        using var cancellationSource = new CancellationTokenSource();
+        cancellationSource.CancelAfter(TimeSpan.FromSeconds(2));
+        await signal.WaitAsync(cancellationSource.Token);
         Assert.True(serializationSuccessful);
```

Comment:
I don't recall seeing this test be flaky, but I think it might indicate a product bug, not a test bug.

I might be remembering wrong, but by validating that DeferredPublishTasks is empty, we're validating that we published immediately, which is the desired behaviour. Removing that assert and adding in the signal means this test will now wait for that deferred publish to happen, but thats not the desired product behaviour. It is better than before when we'd never publish though.

I'll have a look on Monday and see if I can work out what else might be happening.

Reply:
I stepped through the test and its entirely synchronous, but it could fail if xunit orders tests randomly, as there is shared state in the test class. I'll push up a fix that removes that and I think it will be fine after that.

Reply:
Well that didn't work. I have no idea why this is failing in CI.

Reply:
Oh okay, `RunOnDispatcherThreadAsync` doesn't wait for the task to finish, just to be added to the scheduler. On a different machine, the dispatcher could still have notifications to process from previous calls, and maybe we're initializing the publisher before they run, so its responding to an earlier event.